### DISCUSSION TITLE
Data Structures

### DISCUSSION THREAD

**Discussion by porres** [pierreguillot/Camomile#256] [Ideas]
Hi, I'd like to discuss about the possibility of Data Structures support. It is important for GUI desgn.

**Comment by emviveros** [comment]
Yes! What's needed to implement support to Data Structures? What the alternatives to build better graphical interfaces from Pd patches?

**Reply by pierreguillot** [reply]
We need to write code that translates data structure into Juce painting instructions and to add a mechanism to update the interface when it's necessary and also another mechanism that pass mouse events from Juce to the data structure. And if I remember well, the data structure is not the easiest part of Pd code 😅

**Reply by emviveros** [reply]
Ok... So until have Data Structure ported to Juce drawing, nothing to do unless rewrite each interface object in Juce...

**Reply by porres** [reply]
> We need to write code that translates data structure into Juce painting instructions

this is what you do with the other built-in GUI objects then, huh?

**Reply by pierreguillot** [reply]
> Ok... So until have Data Structure ported to Juce drawing, nothing to do unless rewrite each interface object in Juce...

Yes, but it might be simpler than integrating the data structure mechanism. Another solution could be to integrate TCL directly in Camomile but I'm afraid that it becomes even more complicated... Anyway, I'm not convinced that it worth spending time on this feature, I think we can imagine other solutions that would solve the graphical limitation of Camomile such as offering a good Ofelia integration or allowing the use of dynamic images...

**Reply by pierreguillot** [reply]
> > We need to write code that translates data structure into Juce painting instructions
> 
> this is what you do with the other built-in GUI objects then, huh?

Yes but Pd UI are never composed with more than a rectangle, a circle and a line, so that's not very complicated.

**Reply by emviveros** [reply]
> I think we can imagine other solutions that would solve the graphical limitation of Camomile such as offering a good Ofelia integration

Ofelia integration is something I was thinking about... It can provide cool possibilities of interactions for plugins. But once Camomile have implemented the use of pd~ in beta 1.0.8, worth it recompile camomile with ofelia support? Like we've done in Camomile-ELSE?

**Reply by pierreguillot** [reply]
It would be great to use an ofelia window as the main plugin interface (not a floating window in addition to the main patch interface). But perhaps a Camomile-ofelia version is not necessary, if we could specify an "external" window to Camomile?

**Reply by emviveros** [reply]
In that way:
 1. would be needed to rewrite the camomile interface in Ofelia, once open plugin will open ofelia window
 2. rewrite Camomile interface includes Pd gui objects?
 3. if ofelia won't ship with Camomile (it add more dependencies and complicate build process) it will need to add som flag to ofelia compatibility?

What you think about?

**Reply by pierreguillot** [reply]
I'm not sure to understand. My idea is that Ofelia could define an OpenGL context to use and so Camomile wouldn't manage the drawing nor the mouse event anymore, it would just use the context as the plugin graphical interface. I don't know how to do it and even if it's feasible. And perhaps Ofelia is not the solution, perhaps that another scripting language would be more adapted. But the idea is to allow creating a specific graphical interface and not being limited to the Pd GUIs.

**Reply by emviveros** [reply]
I think we are talking the same.. Just that I put more emphasis on the part after OpenGL context defined. That's is something which need some research.. I don't know nothing about Juce interfaces, but I will see it.

About rewriting interfaces, I wanted to say is if we will need to recreate the Camomile Console. It can be solved after discover if Juce GUIs allow specify the internal window part which will comunicate with OpenGL or if the only way is abandon Juce GUI and recreate all Camomile JUCE GUI in OpenGL context window.

About GUIs in GOP patches, Jona is working in something specially usefull! [Basic GUI elements for Ofelia](https://forum.pdpatchrepo.info/topic/13437/some-basic-ofelia-gui-elements)
